python/cuml/manifold/umap.pyx

```diff
@@ -303,156 +303,146 @@ class UMAP(Base):
-        umap_params.local_connectivity = state["local_connectivity"]
-        umap_params.repulsion_strength = state["repulsion_strength"]
-        umap_params.negative_sample_rate = state["negative_sample_rate"]
-        umap_params.transform_queue_size = state["transform_queue_size"]
-        umap_params.init = state["init"]
-        umap_params.a = state["a"]
-        umap_params.b = state["b"]
-        umap_params.target_n_neighbors = state["target_n_neighbors"]
-        umap_params.target_weights = state["target_weights"]
-        umap_params.target_metric = state["target_metric"]
-
-        state["umap_params"] = <size_t> umap_params
+    @staticmethod
+    def _build_umap_params(cls):
+        cdef UMAPParams* umap_params = new UMAPParams()
+        umap_params.n_neighbors = <int> cls.n_neighbors
+        umap_params.n_components = <int> cls.n_components
+        umap_params.n_epochs = <int> cls.n_epochs
+        umap_params.learning_rate = <float> cls.learning_rate
+        umap_params.min_dist = <float> cls.min_dist
+        umap_params.spread = <float> cls.spread
+        umap_params.set_op_mix_ratio = <float> cls.set_op_mix_ratio
+        umap_params.local_connectivity = <float> cls.local_connectivity
+        umap_params.repulsion_strength = <float> cls.repulsion_strength
+        umap_params.negative_sample_rate = <int> cls.negative_sample_rate
+        umap_params.transform_queue_size = <int> cls.transform_queue_size
+        umap_params.verbose = <bool> cls.verbose
+        umap_params.a = <float> cls.a
+        umap_params.b = <float> cls.b
+        if cls.init == "spectral":
+            umap_params.init = <int> 1
+        else:  # self.init == "random"
+            umap_params.init = <int> 0
+        umap_params.target_n_neighbors = <int> cls.target_n_neighbors
+        if cls.target_metric == "euclidean":
+            umap_params.target_metric = MetricType.EUCLIDEAN
+        else:  # self.target_metric == "categorical"
+            umap_params.target_metric = MetricType.CATEGORICAL
+        umap_params.target_weights = <float> cls.target_weights
+        umap_params.random_state = <uint64_t> cls.random_state
+        umap_params.multicore_implem = <bool> cls.multicore_implem
+        umap_params.optim_batch_size = <int> cls.optim_batch_size
 
-        self.__dict__.update(state)
+        cdef uintptr_t callback_ptr = 0
+        if cls.callback:
+            callback_ptr = cls.callback.get_native_callback()
+            umap_params.callback = <GraphBasedDimRedCallback*>callback_ptr
+
+        return <size_t>umap_params
+
+    @staticmethod
+    def _destroy_umap_params(ptr):
+        cdef UMAPParams* umap_params = <UMAPParams*> <size_t> ptr
+        free(umap_params)
 
     @staticmethod
     def _prep_output(X, embedding):
         if isinstance(X, cudf.DataFrame):
-            return cudf.DataFrame.from_gpu_matrix(embedding)
+            return embedding.to_output('dataframe')
```

Comment:
We shouldn't need this function at all anymore because the output type should be getting specified in the constructor (see a few of the other estimators that support `CumlArray` for an example.